substantial/types.py

```diff
@@ -111,17 +141,42 @@ async def exec(self, ctx, save_id, counter) -> Any:
                 ret = op
             else:
                 raise Exception(
-                    f"Expected value or coroutine object, got {type(op)} instead"
+                    f"Expected value or coroutielf.ne object, got {type(op)} instead"
                 )
 
             save = events.Save(save_id, json.dumps(ret), -1)
             ctx.source(events.Event(save=save))
             return ret
         except Exception as e:
+            if self.compensation_stack and len(self.compensation_stack):
+                self.compensation_stack.reverse()
+                for compensation_fn in self.compensation_stack:
+                    try:
+                        compensation_result = compensation_fn()
+                        if inspect.iscoroutine(compensation_result):
+                            compensation_result = await compensation_result
+                        ctx.source(
+                            events.Event(
+                                compensation=events.Compensation(
+                                    save_id=save_id,
+                                    error=str(e),
+                                    compensation_result=json.dumps(
+                                        compensation_result,
+                                    ),
+                                )
+                            )
+                        )
+                    except Exception as compensation_error:
+                        raise CompensationFailed(e, compensation_error)
```

Comment:
You should throw an exception containing the original exception after all compensations are done and not continue. 

Also before you close this up, do you think compensation should happen after all save retries are exhausted or should it happen right after a fail (save fn is executed only once and save retries are ignored)  or something else..? wdyt 🤔

Reply:
Humm, compensate after all save retries are exhausted, But allow user if they  want compensate_on_first_fail, on add it RetryStrategy

Reply:
That's a good idea.